evalutils/stats.py

```diff
@@ -107,19 +107,26 @@ def distance_transform_edt_float32(  # noqa: C901
     # calculate the feature transform
     input = np.atleast_1d(np.where(input, 1, 0).astype(np.int8))
 
-    garbage_collect = gc.collect if input.nbytes > 100e6 else lambda: None
+    def nop():
+        pass
+
+    garbage_collect: Callable[[], Any] = nop
+    if input.nbytes > 100e6:
+        garbage_collect = gc.collect
```

Comment:
This is really ugly. Why not annotate the variable directly without all of this redirection?

Reply:
I agree that it is ugly, MyPy doesn't tolerate:

```python
garbage_collect: Callable = gc.collect if input.nbytes > 100e6 else lambda: None

or

garbage_collect = gc.collect if input.nbytes > 100e6 else lambda: None
```

Then it gives

```
evalutils\stats.py:111: error: Incompatible types in assignment (expression has type "function", variable has type "Callable[..., Any]")
evalutils\stats.py:111: error: Cannot infer type of lambda
```

So I adjusted it to a form that was acceptable for MyPy:

```python
garbage_collect: Callable = lambda: None if input.nbytes <= 100e6 else gc.collect

or

garbage_collect = lambda: None if input.nbytes <= 100e6 else gc.collect
```

But this raises a flake8 error:

` E731 do not assign a lambda expression, use a def`

So we'll introduce the `nop` method to replace the `lambda: None` parts, strangely enough this reintroduces new MyPy errors:

```
Cannot call function of unknown type
```

Or with the added Callable typing:

```
 error: Incompatible types in assignment (expression has type "function", variable has type "Callable[..., Any]")
```

Apparently, this is an issue related to https://github.com/python/mypy/issues/708

**Solution 1:**
Then they suggest using the following as a workaround, which also seems very ugly to me, but fixes the MyPy errors:

```python
garbage_collect: Optional[Callable] = nop if input.nbytes <= 100e6 else gc.collect
assert garbage_collect is not None
```
**Solution 2:**
An alternative would be to just replace the garbage_collect definition with the following. The method is slightly more involved (probably neglectable) than calling the nop and gc.collect methods directly, but it removes the ugly `nop` method:
```python
    def garbage_collect():
        if input.nbytes > 100e6:
            gc.collect()
```
**Solution 3:**
Or to keep it as it is right now:
```python
    garbage_collect: Callable[[], Any] = nop
    if input.nbytes > 100e6:
        garbage_collect = gc.collect
```

**Solution 4:**
Of course, it is also possible to ignore some flake8 or MyPy warnings
Going for ignoring E731 seems like a clean fix here.

Personally, I lean towards either solution 2 or solution 4, what is your preference?

Reply:
Defining a `nop` function to avoid using a `lambda` is dumb. Solution 2 is the least worst solution to me. Why even guard the garbage collection with input size? Is the impact going to be that big if you simply called `gc.collect()` directly as this function is sacrificing compute time for memory anyway.

Reply:
True, I just retained the current implementation. I am also fine with just calling `gc.collect()` everywhere instead of using these constructions. I don't know exactly why the guard was done, because I didn't implement this method (probably to avoid runtime spikes due to gc.collect on small images?). Shall I go for (2) or just call `gc.collect` directly?

Reply:
Nor do I, that's why I asked, you could profile it to find out if it is not in the history. Or just do it directly, if you implement a solution for #303 then I imagine that the strategy will use a method a lot like this check.

Reply:
> Why even guard the garbage collection with input size?

Ideally you would not call gc at all. That was my ad-hoc solution to allow this to work in your absence @jmsmkn , when some study used this method and did not work because of memory constraints. The garbage_collection has significant performance disadvantages, which was why I guarded it, but don't has me where and why.

Ideally, you would not call this at all however, but box the python interpreter into its memory limits using ulimit. That will make this step run automagically. Then we could just remove it. Without ulimits, the function stepped over its --memory OOM-killer boundary when it entered the C-code because some temporary arrays were not cleaned up in time by python garbage collection.

Reply:
Solution 2 does not work and will be buggy! `input.nbytes` changes throughout the function!

Reply:
This would be correct instead.

~~~~~~
is_big = input.nbytes > 100e6
def garbage_collect():
    if big:
        gc.collect()
~~~~~~

Reply:
> Solution 2 does not work and will be buggy! `input.nbytes` changes throughout the function!

Good point I missed that! I have updated this PR with solution (2) using your correction.